lib/tiniergltf/tiniergltf.hpp

```diff
@@ -0,0 +1,1355 @@
+#pragma once
+
+#include <json/json.h>
+#include <functional>
+#include <stack>
+#include <string>
+#include <string_view>
+#include <variant>
+#include <vector>
+#include <array>
+#include <optional>
+#include <limits>
+#include <stdexcept>
+#include <unordered_map>
+#include <unordered_set>
+#include "base64.h"
+
+namespace tiniergltf {
+
+static inline void check(bool cond) {
+	if (!cond)
+		throw std::runtime_error("invalid glTF");
+}
+
+template <typename T>
+static inline void checkIndex(const std::optional<std::vector<T>> &vec,
+		const std::optional<std::size_t> &i) {
+	if (!i.has_value()) return;
+	check(vec.has_value());
+	check(i < vec->size());
+}
+
+template <typename T>
+static inline void checkIndex(const std::vector<T> &vec,
+		const std::optional<std::size_t> &i) {
+	if (!i.has_value()) return;
+	check(i < vec.size());
+}
+
+template <typename T, typename F>
+static inline void checkForall(const std::optional<std::vector<T>> &vec, const F &cond) {
+	if (!vec.has_value())
+		return;
+	for (const T &v : vec.value())
+		cond(v);
+}
+
+template <typename T>
+static inline void checkDuplicateFree(const std::vector<T> &vec) {
+	check(std::unordered_set<T>(vec.begin(), vec.end()).size() == vec.size());
+}
+
+template <typename T>
+static inline T as(const Json::Value &o);
+
+template<>
+bool as(const Json::Value &o) {
+	check(o.isBool());
+	return o.asBool();
+}
+
+template<>
+double as (const Json::Value &o) {
+	check(o.isDouble());
+	return o.asDouble();
+}
+
+template<>
+std::size_t as(const Json::Value &o) {
+	check(o.isUInt64());
+	auto u = o.asUInt64();
+	check(u <= std::numeric_limits<std::size_t>::max());
+	return u;
+}
+
+template<>
+std::string as(const Json::Value &o) {
+	check(o.isString());
+	return o.asString();
+}
+
+template<typename U>
+std::vector<U> asVec(const Json::Value &o) {
+	check(o.isArray());
+	std::vector<U> res;
+	res.reserve(o.size());
+	for (Json::ArrayIndex i = 0; i < o.size(); ++i) {
+		res.push_back(as<U>(o[i]));
+	}
+	return res;
+}
+
+template<typename U, std::size_t n>
+std::array<U, n> asArr(const Json::Value &o) {
+	check(o.isArray());
+	check(o.size() == n);
+	std::array<U, n> res;
+	for (Json::ArrayIndex i = 0; i < n; ++i) {
+		res[i] = as<U>(o[i]);
+	}
+	return res;
+}
+
+struct AccessorSparseIndices {
+	std::size_t bufferView;
+	std::size_t byteOffset;
+	enum class ComponentType {
+		UNSIGNED_BYTE,
+		UNSIGNED_SHORT,
+		UNSIGNED_INT,
```

Comment:
According to the spec, short is >= 2 bytes, and int is also >= 2 bytes. Perhaps `S16` and `S32` would be more accurate names - unless this is supposed to be platform-specific.

Same comment for `Accessor`.

Reply:
The C(++) spec(s), yes. But OpenGL and glTF use these names more sensibly, and I prefer to be consistent with those. See e.g. https://www.khronos.org/opengl/wiki/OpenGL_Type and https://registry.khronos.org/glTF/specs/2.0/glTF-2.0.html#accessor-data-types. I can add a comment.